### ISSUE TITLE
PMML model makes wrong predictions after PMML pipeline is converted to a PMML file

### ISSUE THREAD

**Issue by yj333** [jpmml/sklearn2pmml#311]
Hi,
We are working on a regression model using XGBoost regressor. The predictions predicted by the PMML pipeline look reasonable after the pipeline was fitted on the training data. However, after the pipeline was converted to a PMML file, the predictions generated by the PMML model (imported through PyPMML) are two to three order-of-magnitude bigger. Strangely enough, this issue did not happen when we switched the entire training data with a small subset of data.

The entire training set has tens of millions of records. There are about 10 features in the training data - majority being categorical. Our numerical features are specified as int or float and all the categorical features are strings.  Here are the code to create and fit the pipeline.

```
def pmml_pipeline(num_features, cat_features):
    reg_mod = XGBRegressor(n_estimators = 300,\
                         max_depth = 10,\
                         learning_rate = 0.3,\
                         subsample = 1,\
                         colsample_bytree = 1,\
                         gamma = 1,
                         objective = ‘reg:squarederror’,
                         n_jobs=-1)
    categorical_pipe = Pipeline([(“imputer”, SimpleImputer(strategy=“most_frequent”)),\
                           (“ohe”, OneHotEncoder(handle_unknown=“ignore”))])
    numerical_pipe = Pipeline([(“imputer”, SimpleImputer(strategy=“mean”)),\
                               (“scaler”,StandardScaler())])
    preprocessor = ColumnTransformer(
        transformers=
        [
            (“cat”, categorical_pipe, cat_features),
            (“num”, numerical_pipe, num_features),
        ]
    )
    p = PMMLPipeline(
        [
            (“preprocessor”, preprocessor),
            (“regressor”, reg_mod),
        ])
    return p
ppln = pmml_pipeline()
ppln.fit(X_train,y_train)
sklearn2pmml(ppln, ‘myPMML.pmml’, with_repr = True)
# predictions directly from pipeline
y_pred = ppln.predict(X_test)
# predictions from PMML model by PyPMML
from pypmml import Model
model = Model.fromFile(‘myPMML.pmml’)
y_pred_pmml = model.predict(X_test)
```

The issues here are, for example, y_pred = 60 but the corresponded y_pred_pmml = 3000.  Using JPMML Evaluator produces the same issue. Both X_train and X_test are pandas Dataframe, and y_train is a pandas Series. The above pipeline uses ColumnTransformer to transform features.  However, if DataFrameMapper is used then the issue seems to be gone but it takes much more memory during training and thus becomes unfeasible for us. It seems that something is wrong at the step when the pipeline with ColumnTransformer was converted to a PMML. Do you spot anything that is inappropriate? Thank you!

**Comment by vruusmann** [comment]
> OneHotEncoder(handle_unknown=“ignore”)

You can't use `OneHotEncoder` for encoding categorical features when the final estimator is some XGBoost estimator.

The reason being is that `OneHotEncoder` produces sparse data frames, which mis-represents the status of individual category levels (should be `1` for set, and `0` for unset; not `1` for set and `NaN` for unset).

Your whole Scikit-Learn + XGBoost is making non-sensical predictions in Python.

> However, if DataFrameMapper is used then the issue seems to be gone

This is so, because `DataFrameMapper` transforms sparse data matrices to dense matrices.

Whn using `ColumnTransformer` you should be specifying `OneHotEncoder(sparse = False)`. Then you shall be getting the same results as with `DataFrameMapper`, and Python and PMML predictions will also agree.

> from pypmml import Model
> model = Model.fromFile(‘myPMML.pmml’)

Any particular reason for choosing PyPMML over [JPMML-Evaluator-Python](https://github.com/jpmml/jpmml-evaluator-python)?

JPMML-Evaluator-Python is always correct, and is anywhere between 10 -- 100x more performant.

**Comment by yj333** [comment]
Thank you for your suggestions!  

Setting `OneHotEncoder(sparse = False)` with `ColumnTransformer` does help in our case - now PMML predictions seem to make much more sense.  The only concern is having a dense matrix just like what `DataFrameMapper` is able to generate is really memory expensive.

But there is still an issue - the PMML model generates an error when there are missing data in the test set. Using `SimpleImputer()` with `ColumnTransformer` does not seem to be helpful when there is NaN value in the test set.  The error from JPMML Evaluator says something like `org.jpmml.evaluator.ValueCheckException: Field "x1" cannot accept invalid value NaN`.  Again, if `SimpleImputer()` is used with `DataFrameMapper` then the PMML model can handle NaN value correctly.  As mentioned above, the only issue with `DataFrameMapper` is it takes too much memory to train the model.  Do you know why this is and is there a remedy to it?

I saw that there are previous discussion on this before [#168](https://github.com/jpmml/jpmml-sklearn/issues/168) so I tried to do something similar as below.

```
cat_pipe = Pipeline([
	('domain', CategoricalDomain(invalid_value_treatment = "as_value", invalid_value_replacement = "Other")),('ohe',OneHotEncoder(handle_unknown="ignore",sparse=False))
	])
    
num_pipe = Pipeline([
	('domain', ContinuousDomain(missing_value_treatment = "as_mean",invalid_value_treatment = "as_value", invalid_value_replacement = 300)),('scaler',StandardScaler())
	])

preprocessor = ColumnTransformer([ 
      ("cat", cat_pipe, cat_features), 
      ("num", num_pipe, num_features) 
	])
ppln = PMMLPipeline(
        [
            ("preprocessor", preprocessor),
            ("regressor",XGBRegressor()),
        ])
ppln.fit(X=X_train,y=y_train)
```

Q1: I got an error saying `ValueError: y should be a 1d array, got an array of shape (m,n)` after I fit the pipeline where m = length of y_train and n = number of categorical features. It appears the `OneHotEncoder` transforms the target variables as well when the domain decorator is included. Not sure why.

Q2: So I removed the domain decorator from the categorical feature pipeline and just have the following

```
cat_pipe = Pipeline([
        ('ohe',OneHotEncoder(handle_unknown="ignore",sparse=False))
    ])
```
It worked and the PMML predictions also seemed fine. But it seems `missing_value_treatment = "as_mean"` is not doing its job - when I replaced the NaN with 300 in one of the numerical features, the two predictions were the same.  It appears the pipeline treats all NaN value as invalid and replaces those with 300 as specified in the pipeline.  When I checked the document [here](http://dmg.org/pmml/v4-4-1/MiningSchema.html) it does state that `missingValueTreatment` is just for information only - the user needs to specify the mean value if replacing missing values with the mean is needed. Just want to confirm if I understand it correctly. Thanks!

**Comment by vruusmann** [comment]
>  The only concern is having a dense matrix is really memory expensive.

The choice is between "memory expensive and correct" and "memory cheap and incorrect" :-)

You were assuming that `OneHotEncoder(sparse = True)` and `OneHotEncoder(sparse = False)` would be giving you identical models (identical in the inferred predictive function sense). This assumption holds true with many ML model algorithms, but **_not_** with the XGBoost algorithm.

The `DataFrameMapper` approach sides with `OneHotEncoder(sparse = False`). So, it's two against one.

> Using `SimpleImputer()` does not seem to be helpful when there is NaN value in the test set.

The XGBoost algorithm is able to deal with missing values natively both during model training and prediction (aka testing) time.

You shouldn't be using an imputer step at all - it disfigures missing values, which skews feature distributions, and leads to a "stupider" XGBoost model in the end.

In newer Scikit-Learn versions the `OneHotEncoder` transformer should be able to pass through missing values (in older versions it breaks). Alternatively, consider using a missing value-aware OHE transformer class `category_encoders.one_hot.OneHotEncoder` from the Category-Encoders package.

The SkLearn2PMML package should be able to deal with any OHE configuration (with or without missing values, with or without invalid values). Some of my integration tests:
https://github.com/jpmml/jpmml-sklearn/blob/1.7.0/pmml-sklearn-extension/src/test/resources/extensions/category_encoders.py

> Q1: I got an error saying ValueError: `y should be a 1d array, got an array of shape (m,n)`

Use the `sklearn2pmml.util.Reshaper` meta-transformer to reshape the data matrix between steps:

```python
mapper = DataFrameMapper([
  (["col"], [FirstTransformer(), Reshaper(newshape = (-1, 1)), SecondTransformer()]
])
```

> Q2: it seems `missing_value_treatment = "as_mean"` is not doing its job

The `Domain.missing_value_treatment` attribute must be used together with the `Domain.missing_value_replacement` attribute - the former declares the name of the method, the latter provides the actual value that was obtained by applying the method.

In other words, you'd need to calculate the mean value yourself, and then do:
```python
domain = ContinuousDomain(missing_value_treatment = "as_mean", missing_value_replacement = 300)
```

Using the `Domain.missing_value_treatment` attribute in isolation should raise a conversion error.

**Comment by vruusmann** [comment]
> the only issue with DataFrameMapper is it takes too much memory to train the model. Do you know why this is and is there a remedy to it?

This problem can only be solved by activating XGBoost "native categorical feature" support:
https://xgboost.readthedocs.io/en/latest/python/examples/categorical.html

It seems to be in a preview stage in XGBoost 1.5.X. Once it stabilizes enough, and becomes available for CPU executors, I intend to implement it into the JPMML-XGBoost library (and then propagate downstream into JPMML-SkLearn, JPMML-R, JPMML-SparkML etc. libraries).

**Comment by vruusmann** [comment]
> This problem can only be solved by activating XGBoost "native categorical feature" support:
> https://xgboost.readthedocs.io/en/latest/python/examples/categorical.html

XGBoost categorical feature support is available starting from SkLearn2PMML version 0.79.0.

See https://openscoring.io/blog/2022/04/12/onehot_encoding_sklearn_xgboost_pipeline/
